lib/spack/spack/cmd/update.py

```diff
@@ -0,0 +1,102 @@
+# Copyright 2013-2023 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+import re
+import sys
+from argparse import ArgumentParser, Namespace
+from typing import Dict, List
+
+from llnl.util import tty
+from llnl.util.filesystem import working_dir
+from llnl.util.tty.colify import colify
+
+import spack.cmd
+import spack.util.git
+from spack.cmd import spack_is_git_repo
+
+description = "update the spack repository to the latest commit"
+section = "system"
+level = "short"
+
+
+def setup_parser(subparser: ArgumentParser) -> None:
+    subparser.add_argument("-b", "--branch", help="name of the branch to upate the repository to")
```

Comment:
```suggestion
    subparser.add_argument("-b", "--branch", help="name of a branch to update the repository to")
```

but I think this would be better as:

```suggestion
    mutex = subparser.add_mutually_exclusive_group()
    mutex.add_argument("release_or_branch", nargs="?", "--help="name of a branch or release to use (optional)")
    
    mutex.add_argument("-l", "--list-releases", "--help="list names of known spack releases")
    mutex.add_argument("--list-branches", "--help="list names of all known branches")
```

where:
* `spack update` pulls the current branch
* `spack update <ref>` checks out the given release or branch (they're the same but I think we should just show releases by default)
* and the two list options list either just release branches or all branches

Reply:
I'm game for `release_or_branch` and `-l` but I'm not sure how useful `--list-branches` will be unless we stop encouraging folks from creating branches directly in the Spack repository as we now have ~125 branches.